python/revng/ptml/tree.py

```diff
@@ -0,0 +1,99 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+# noqa: A003
+
+from dataclasses import dataclass
+from functools import cached_property
+from typing import Dict, Generator, List
+
+
+@dataclass
+class Position:
+    """Represents a single character in a multi-line text document"""
+
+    line: int
+    character: int
+
+    def clone(self):
+        return Position(self.line, self.character)
+
+    def to_string(self):
+        return f"{self.line}:{self.character}"
+
+
+@dataclass
+class Range:
+    """Represents a range of characters in a multi-line text document. Both
+    start and end are inclusive"""
+
+    start: Position
+    end: Position
+
+    def is_single_line(self):
+        return self.start.line == self.end.line
+
+    def to_string(self):
+        return f"{self.start.to_string()},{self.end.to_string()}"
+
+
+class TextExtractor:
+    """Utility function that allows extracting `Range`s from a text"""
+
+    def __init__(self, text: str):
+        self.original_text = text
+
+    @cached_property
```

Comment:
`@cached_property` has backfired in a number of situations in the past.

Please do the thing by hand, so that the cache it doesn't go unnoticed in hypothetical future changes.

Reply:
> @cached_property has backfired in a number of situations in the past.

This has always come across as superstitious, can you point to an example where the use of `cached_property` was problematic? Doing things by hand is usually more error prone for gaining nothing compared to the decorator.

Reply:
This has happened multiple times with @fcremo, to the point where it became a meme, but I agree that it has transcended into superstition.

At least, please add a comment on `self.original_text` definition about the existence of a cache.

Reply:
Added a comment